consensus/polybft/validator_set.go

```diff
@@ -18,46 +77,365 @@ type ValidatorSet interface {
+	var validators = make([]*ValidatorAccount, len(valz))
+	for i, v := range valz {
+		validators[i] = NewValidator(v, 0)
+	}
+
+	validatorSet := &validatorSet{
+		validators: validators,
+	}
+
+	err := validatorSet.updateWithChangeSet()
+	if err != nil {
+		return nil, fmt.Errorf("cannot update changeset: %w", err)
+	}
+
+	return validatorSet, nil
+}
+
+// IncrementProposerPriority increments ProposerPriority of each validator and
+// updates the proposer.
+func (v *validatorSet) IncrementProposerPriority(times uint64) error {
+	if v.isNilOrEmpty() {
+		return fmt.Errorf("validator set cannot be nul or empty")
+	}
+
+	if times == 0 {
+		return fmt.Errorf("cannot call IncrementProposerPriority with non-positive times")
+	}
+
+	// Cap the difference between priorities to be proportional to 2*totalPower by
+	// re-normalizing priorities, i.e., rescale all priorities by multiplying with:
+	//  2*totalVotingPower/(maxPriority - minPriority)
+	vp, err := v.TotalVotingPower()
+
+	if err != nil {
+		return fmt.Errorf("cannot calculate total voting power")
+	}
+
+	diffMax := PriorityWindowSizeFactor * vp
+
+	err = v.rescalePriorities(diffMax)
+	if err != nil {
+		return fmt.Errorf("cannot rescale priorities: %w", err)
+	}
+
+	err = v.shiftByAvgProposerPriority()
+	if err != nil {
+		return fmt.Errorf("cannot shift avg priorities: %w", err)
+	}
+
+	var proposer *ValidatorAccount
+
+	for i := uint64(0); i < times; i++ {
+		proposer, err = v.incrementProposerPriority()
+		if err != nil {
+			return fmt.Errorf("cannot increment proposer priority: %w", err)
+		}
+	}
+
+	v.proposer = proposer
+
+	return nil
+}
+
+func (v *validatorSet) incrementProposerPriority() (*ValidatorAccount, error) {
+	for _, val := range v.validators {
+		// Check for overflow for sum.
+		newPrio := safeAddClip(val.ProposerPriority, int64(val.Metadata.VotingPower))
+		val.ProposerPriority = newPrio
+	}
+	// Decrement the validator with most ProposerPriority.
+	mostest, err := v.getValWithMostPriority()
+	if err != nil {
+		return nil, fmt.Errorf("cannot get validator with most priority: %w", err)
+	}
+	// Mind the underflow.
+	vp, err := v.TotalVotingPower()
+	if err != nil {
+		return nil, fmt.Errorf("cannot get total voting power: %w", err)
+	}
+
+	mostest.ProposerPriority = safeSubClip(mostest.ProposerPriority, vp)
+
+	return mostest, nil
+}
+
+// TotalVotingPower returns the sum of the voting powers of all validators.
+// It recomputes the total voting power if required.
+func (v *validatorSet) TotalVotingPower() (int64, error) {
+	if v.totalVotingPower == 0 {
+		err := v.updateTotalVotingPower()
+		if err != nil {
+			return 0, fmt.Errorf("cannot update total voting power: %w", err)
+		}
+	}
+
+	return v.totalVotingPower, nil
+}
+
+func (v *validatorSet) updateWithChangeSet() error {
+	err := v.updateTotalVotingPower()
+	if err != nil {
+		return fmt.Errorf("cannot update total voting power: %w", err)
+	}
+	// Scale and center.
+	totalVotingPower, err := v.TotalVotingPower()
```

Comment:
Do we need this `TotalVotingPower` call here, since `updateTotalVotingPower` recalculates it. I think it is enough to call `v.totalVotingPower`.

Reply:
good catch, it is not necessary since it will be calculate